src/repr/type_binary.ml

```diff
@@ -18,6 +18,311 @@ open Type_core
+    | Product of t list
+    | Variant of t list
+    (* Custom codecs and bijections must be versioned with UUIDs. *)
+    | Opaque of Uuid.t
+    | Map of Uuid.t * t
+    (* Recursion represented via De Bruijn indexing: [Recur n] indicates a
+       recursive occurrence of the [n]-th [Recursive] node above it in the
+       shape. *)
+    | Recursive of t
+    | Recur of int
+
+  (** TODO: reorganise libraries so that we can use [ppx_repr] here *)
+  let len_t =
+    let open Type_combinators in
+    enum "len"
+      [
+        ("int", `Int);
+        ("int8", `Int8);
+        ("int16", `Int16);
+        ("int32", `Int32);
+        ("int64", `Int64);
+      ]
+
+  let size_t =
+    let open Type_combinators in
+    variant "size" (fun boxed fixed any -> function
+      | Boxed x -> boxed x | Fixed x -> fixed x | Any -> any)
+    |~ case1 "boxed" len_t (fun x -> Boxed x)
+    |~ case1 "fixed" int (fun x -> Fixed x)
+    |~ case0 "any" Any
+    |> sealv
+
+  let t =
+    let open Type_combinators in
+    mu (fun t ->
+        variant "shape"
+          (fun
+            empty
+            bool
+            char
+            int
+            int32
+            int64
+            float
+            option
+            contiguous
+            product
+            variant
+            opaque
+            map
+            recursive
+            recur
+          -> function
+          | Empty -> empty
+          | Bool -> bool
+          | Char -> char
+          | Int -> int
+          | Int32 -> int32
+          | Int64 -> int64
+          | Float -> float
+          | Option x -> option x
+          | Contiguous (a, b) -> contiguous (a, b)
+          | Product x -> product x
+          | Variant x -> variant x
+          | Opaque x -> opaque x
+          | Map (a, b) -> map (a, b)
+          | Recursive x -> recursive x
+          | Recur x -> recur x)
+        |~ case0 "empty" Empty
+        |~ case0 "bool" Bool
+        |~ case0 "char" Char
+        |~ case0 "int" Int
+        |~ case0 "int32" Int32
+        |~ case0 "int64" Int64
+        |~ case0 "float" Float
+        |~ case1 "option" t (fun x -> Option x)
+        |~ case1 "contiguous" (pair size_t t) (fun (a, b) -> Contiguous (a, b))
+        |~ case1 "product" (list t) (fun x -> Product x)
+        |~ case1 "variant" (list t) (fun x -> Variant x)
+        |~ case1 "opaque" Uuid.t (fun x -> Opaque x)
+        |~ case1 "map" (pair Uuid.t t) (fun (a, b) -> Map (a, b))
+        |~ case1 "recursive" t (fun x -> Recursive x)
+        |~ case1 "recur" int (fun x -> Recur x)
+        |> sealv)
+
+  let shape_equal = unstage (Type_ordered.equal t)
+
+  (** Deriving shapes from type reps is straightforward, but requires some care
+      with recursive types. We convert recursive loops to use De Bruijn indexing
+      by unfolding each one with a fresh placeholder and tracking recursion
+      depth so that we can replace the placeholders with the right index.
```

Comment:
I don't understand all the magic yet, but the code is interesting to read!

Reply:
I'll try and make it clearer on a subsequent pass :slightly_smiling_face:

Reply:
Thank you for all your clarifications!